rust/protocol/src/error.rs

```diff
@@ -1,198 +1,131 @@
 //
-// Copyright 2020 Signal Messenger, LLC.
+// Copyright 2020-2021 Signal Messenger, LLC.
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
 use crate::curve::KeyType;
 
-use std::error::Error;
+use displaydoc::Display;
+use thiserror::Error;
+
 use std::fmt;
 use std::panic::UnwindSafe;
 
 pub type Result<T> = std::result::Result<T, SignalProtocolError>;
 
+/// Wraps a boxed error struct and delegates the [std::error::Error] trait to it.
+///
+/// Unlike a `Box<E: Error>` wrapping a concrete error type, a `Box<dyn Error>` wrapping
+/// a type-erased trait object does *not* implement [std::error::Error] itself (see [this
+/// stackoverflow answer] for details). This breaks the ability to nicely use the `#[source]`
+/// annotation from the [thiserror::Error] derive macro. In order to avoid manually impling
+/// [std::error::Error::source] for each case of [SignalProtocolError], we create this wrapper
+/// struct to delegate to the underlying `dyn Error` in order to nicely derive [thiserror::Error]
+/// for [SignalProtocolError::ApplicationCallbackError].
+///
+/// [this stackoverflow answer]: https://stackoverflow.com/a/62452516/2518889
```

Comment:
I don't think the resulting code is clearer than just implementing `source` manually (and therefore Error). In the choice between making one place a lot more complicated and many places a little more complicated, I'd usually err on the side of containing the complexity. ([If only we didn't need UnwindSafe here.](https://github.com/dtolnay/thiserror/pull/41))